processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -1038,4 +1182,398 @@ public void clear()
       facts.clear();
     }
   }
+
+  public static class OnHeapAggregateProjection implements IncrementalIndexRowSelector
```

Comment:
A curiosity question: the name of this seems to be specific to a Projection.  But, like, you could argue that even the "primary" could also be a projection of the incoming data, which *could* show up in the code as having one more projection for the primary along with the other defined projections.  Instead, you've taken the approach of having a different Projection class than the main IncrementalIndex.  Is there a specific reason why *not* modeling the "primary" as just-another-projection ended up being the favored approach?

Reply:
yea, i was (and am still) considering this, especially since `IncrementalIndex` itself is kind of a strange abstraction in the first place and I think it could let us clean some stuff up. The main reason I have not tackled this yet is because it would require being able to represent no rollup projections, which is left out for now, the other was that it would be a bit more dramatic of a change for `IncrementalIndex`, and I was trying to be conservative on overall changes for this PR. I do think it would be a lot nicer though if it was just all projections, so I plan to keep thinking about this.